extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientsImpl.java

```diff
@@ -1,28 +1,80 @@
 package io.quarkus.oidc.client.runtime;
 
+import static io.quarkus.oidc.client.runtime.OidcClientRecorder.DEFAULT_OIDC_CLIENT_ID;
+
 import java.io.Closeable;
 import java.io.IOException;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
-import java.util.function.Function;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import jakarta.annotation.PreDestroy;
+import jakarta.enterprise.inject.Produces;
+import jakarta.inject.Inject;
+import jakarta.inject.Singleton;
+
+import org.jboss.logging.Logger;
 
+import io.quarkus.arc.ClientProxy;
+import io.quarkus.arc.Unremovable;
 import io.quarkus.oidc.client.OidcClient;
 import io.quarkus.oidc.client.OidcClientException;
 import io.quarkus.oidc.client.OidcClients;
+import io.quarkus.oidc.client.Tokens;
+import io.quarkus.oidc.common.OidcEndpoint;
+import io.quarkus.oidc.common.OidcRequestContextProperties;
+import io.quarkus.oidc.common.OidcRequestFilter;
+import io.quarkus.oidc.common.OidcResponseFilter;
+import io.quarkus.oidc.common.runtime.OidcCommonUtils;
+import io.quarkus.oidc.common.runtime.OidcConstants;
+import io.quarkus.oidc.common.runtime.OidcTlsSupport;
+import io.quarkus.runtime.Shutdown;
+import io.quarkus.runtime.configuration.ConfigurationException;
+import io.quarkus.tls.TlsConfigurationRegistry;
 import io.smallrye.mutiny.Uni;
+import io.vertx.core.Vertx;
+import io.vertx.ext.web.client.WebClientOptions;
+import io.vertx.mutiny.core.MultiMap;
+import io.vertx.mutiny.ext.web.client.WebClient;
+
+@Singleton
+public final class OidcClientsImpl implements OidcClients, Closeable {
+
+    private static final Logger LOG = Logger.getLogger(OidcClientsImpl.class);
+    private static final String CLIENT_ID_ATTRIBUTE = "client-id";
 
-public class OidcClientsImpl implements OidcClients, Closeable {
-    private OidcClient defaultClient;
-    private Map<String, OidcClient> staticOidcClients;
-    private Function<OidcClientConfig, Uni<OidcClient>> dynamicOidcClients;
+    private final OidcClient defaultClient;
+    private final Map<String, OidcClient> staticOidcClients;
+    private final Vertx vertx;
+    private final OidcTlsSupport tlsSupport;
+    private volatile PeriodicTask tokenRefreshTask;
```

Comment:
Can we keep `PeriodicTask` only in `OidcClientsImpl` ?

Reply:
Yep